### ISSUE TITLE
Introduce affiliate_wp_customers DB table

### ISSUE THREAD

**Issue by amdrew** [AffiliateWP/AffiliateWP#2491]
PR: #2604

We should introduce a `affiliate_wp_customers` DB table.

Possible DB columns:

- Customer ID
- User ID
- Affiliate ID
- First Name
- Last Name
- Email Address (primary/first email used)
- Date Created
- IP Address (Primary/first used by customer)

In addition, we would store additional email addresses and IP addresses used by the customer in affiliate meta.

The primary advantage of this new table is allowing the site admin to see the relationships between affiliates and customers. A site admin could see how many customers an affiliate has referred, who they are, and more.

Other add-ons such as Lifetime Commissions, Order Details for Affiliates, Signup Referrals could then leverage this data to add additional functionality.

Things needed:

- [x] New database table
- [x] Upgrade routine
- [x] Additional email addresses and IP addresses stored in the affiliate meta

**Comment by pippinsplugins** [comment]
I have this pretty well working. Todo still:

1. Create upgrade routine to populate past customer records
2. Remove the temporary UI I added (there will be no UI for now)
3. Add customer tracking to all integrations
4. Add method to AFFWP_Customer for `get_original_affiliate()`
5. Re-build [Lifetime Commissions](https://github.com/AffiliateWP/affiliate-wp-lifetime-commissions/) to use this new customers table as an initial test to see if it does what we need

**Comment by pippinsplugins** [comment]
First integration of the upgrade routine is added. Note, it won't remove the upgrade notice after it's run. I'll add that once it's solid.

**Comment by pippinsplugins** [comment]
Todo:

- [ ] Add unit tests
- [ ] Add meta query join for querying customers by affiliate ID
- [ ] Finish all integrations
- [ ] Test customer creation for all integrations
- [ ] Test upgrade routine with referrals stored from inactive integration plugins
- [ ] Added better descriptive text to upgrade routine

**Comment by pippinsplugins** [comment]
Todo:

- [ ] Add unit tests
- [ ] Add meta query join for querying customers by affiliate ID
- [ ] Finish all integrations
- [ ] Test customer creation for all integrations
- [ ] Test upgrade routine with referrals stored from inactive integration plugins
- [ ] Added better descriptive text to upgrade routine

**Comment by pippinsplugins** [comment]
Integration tests:

- [ ] Contact Form 7
- [ ] Easy Digital Downloads
- [ ] Caldera Forms
- [ ] Formidable Pro
- [ ] Give
- [ ] Gravity Forms
- [ ] iThemes Exchange
- [ ] Jigoshop
- [ ] LifterLMS
- [ ] MarketPress
- [ ] MemberMouse
- [ ] MemberPress
- [ ] Ninja Forms
- [ ] OptimizeMember
- [ ] PayPal Buttons
- [ ] Paid Memberships Pro
- [ ] Paid Member Subscriptions
- [ ] Restrict Content Pro
- [ ] s2Member
- [ ] Shopp
- [ ] Sprout Invoices
- [ ] Stripe Checkout (through WP Simple Pay)
- [ ] WooCommerce
- [ ] WP EasyCart
- [ ] WP eCommerce
- [ ] WPForms
- [ ] WP-Invoice
- [ ] Zippy Courses

**Comment by pippinsplugins** [comment]
Ready for testing on all integrations but the four above that are unchecked.

**Comment by pippinsplugins** [comment]
I've decided not to add compatibility for the following integrations:

- Contact Form 7
- WP-Invoice
- Sprout Invoices

Customers will still be tracked through those when logged in, so it's not completely lacking for them.

**Comment by pippinsplugins** [comment]
@tubiz could you run through a series of tests on this? If anything is unclear, let me know and I can add further information.

**Comment by tubiz** [comment]
@pippinsplugins 👍

**Comment by tubiz** [comment]
@pippinsplugins what other tests do I need to run apart from testing all integrations to make sure they all work perfectly.

**Comment by tubiz** [comment]
@pippinsplugins with the test I have done so far the customer records are being created properly but the referral record is not created.

In the upgrade routine we need to add a `customer_id` field to the `affiliate_wp_referrals` table as the `customer_id` is now part of the referral data

**Comment by pippinsplugins** [comment]
@tubiz could you please submit a PR for https://github.com/tubiz/AffiliateWP/commit/9dbed4707273f84aa08e970105daec8efb789c79?

**Comment by tubiz** [comment]
@pippinsplugins PR submitted.

PR: #2605

**Comment by pippinsplugins** [comment]
Thanks!

I've brought this branch up to date with `release/2.2` and updated the upgrade routine to create the `customer_id` column in the referrals table.

Ready for additional testing.

**Comment by pippinsplugins** [comment]
Full PR is here https://github.com/AffiliateWP/AffiliateWP/pull/2604/

**Comment by tubiz** [comment]
Okay. I will carry out additional test.

**Comment by pippinsplugins** [comment]
I've found a conflict between the new customer meta table I'd added and EDD's customer meta. Working on resolving it.

**Comment by pippinsplugins** [comment]
Conflict should be fixed. Had to rename the customermeta table and the meta type from `customer` to `affwp_customer`.

To test the fix, please run `affiliate_wp_install()` and then re-run the batch upgrade routine for migrating customers.

**Comment by amdrew** [comment]
Re-ran the upgrade routine and install script for https://github.com/AffiliateWP/affiliate-wp-lifetime-commissions/issues/75.

I saw a bunch of these notices after the upgrade:

```
[15-Mar-2018 04:18:24 UTC] PHP Notice:  Array to string conversion in /Users/andrew/Dropbox/development/sites/affwp/wordpress/wp-includes/formatting.php on line 1045
[15-Mar-2018 04:18:24 UTC] PHP Stack trace:
[15-Mar-2018 04:18:24 UTC] PHP   1. {main}() /Users/andrew/Dropbox/development/sites/affwp/wordpress/wp-admin/admin-ajax.php:0
[15-Mar-2018 04:18:24 UTC] PHP   2. do_action() /Users/andrew/Dropbox/development/sites/affwp/wordpress/wp-admin/admin-ajax.php:97
[15-Mar-2018 04:18:24 UTC] PHP   3. WP_Hook->do_action() /Users/andrew/Dropbox/development/sites/affwp/wordpress/wp-includes/plugin.php:453
[15-Mar-2018 04:18:24 UTC] PHP   4. WP_Hook->apply_filters() /Users/andrew/Dropbox/development/sites/affwp/wordpress/wp-includes/class-wp-hook.php:310
[15-Mar-2018 04:18:24 UTC] PHP   5. affwp_process_batch_request() /Users/andrew/Dropbox/development/sites/affwp/wordpress/wp-includes/class-wp-hook.php:286
[15-Mar-2018 04:18:24 UTC] PHP   6. AffWP\Utils\Batch_Process\Upgrade_Create_Customers->process_step() /Users/andrew/Dropbox/development/sites/affwp/wordpress/wp-content/plugins/AffiliateWP/includes/admin/ajax-actions.php:180
[15-Mar-2018 04:18:24 UTC] PHP   7. AffWP\Referral->set() /Users/andrew/Dropbox/development/sites/affwp/wordpress/wp-content/plugins/AffiliateWP/includes/admin/tools/upgrades/class-batch-upgrade-create-customers.php:128
[15-Mar-2018 04:18:24 UTC] PHP   8. AffWP\Referral->save() /Users/andrew/Dropbox/development/sites/affwp/wordpress/wp-content/plugins/AffiliateWP/includes/abstracts/class-affwp-object.php:173
[15-Mar-2018 04:18:24 UTC] PHP   9. Affiliate_WP_Referrals_DB->update_referral() /Users/andrew/Dropbox/development/sites/affwp/wordpress/wp-content/plugins/AffiliateWP/includes/abstracts/class-affwp-object.php:192
[15-Mar-2018 04:18:24 UTC] PHP  10. sanitize_text_field() /Users/andrew/Dropbox/development/sites/affwp/wordpress/wp-content/plugins/AffiliateWP/includes/class-referrals-db.php:215
[15-Mar-2018 04:18:24 UTC] PHP  11. _sanitize_text_fields() /Users/andrew/Dropbox/development/sites/affwp/wordpress/wp-includes/formatting.php:4697
[15-Mar-2018 04:18:24 UTC] PHP  12. wp_check_invalid_utf8() /Users/andrew/Dropbox/development/sites/affwp/wordpress/wp-includes/formatting.php:4749
[15-Mar-2018 04:18:28 UTC] PHP Notice:  Array to string conversion in /Users/andrew/Dropbox/development/sites/affwp/wordpress/wp-includes/formatting.php on line 1045
```

**Comment by tubiz** [comment]
@amdrew I am unable to replicate this.

**Comment by tubiz** [comment]
PR: #2606 

Fixes

* The upgrade routine can cause a fatal error if an integration is enabled but the plugin is not activated https://github.com/AffiliateWP/affiliate-wp-lifetime-commissions/issues/75
* Referrals not created for a lifetime customer `get_canonical_affiliate_id()`

**Comment by pippinsplugins** [comment]
@DrewAPicture most of your items have been addressed. A few left to do still.

**Comment by pippinsplugins** [comment]
Ok all items addressed.

**Comment by DrewAPicture** [comment]
@pippinsplugins Looks good to me. Pushed a few small adjustments.

**Comment by pippinsplugins** [comment]
Thanks @DrewAPicture!

I'm going to merge this into `release/2.2`!

**Comment by gspice** [comment]
Testing on a woo site, an uncaught exception occurred while updating the database.  To reproduce, delete a WooCommerce order that is referenced by an existing Referral. 

```
[29-Mar-2018 21:13:21 UTC] PHP Fatal error:  Uncaught Exception: Invalid order. in /app/public/wp-content/plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php:102
Stack trace:
#0 /app/public/wp-content/plugins/woocommerce/includes/class-wc-data-store.php(149): Abstract_WC_Order_Data_Store_CPT->read(Object(WC_Order))
#1 /app/public/wp-content/plugins/woocommerce/includes/abstracts/abstract-wc-order.php(109): WC_Data_Store->read(Object(WC_Order))
#2 /app/public/wp-content/plugins/AffiliateWP-issue-2491/includes/integrations/class-woocommerce.php(365): WC_Abstract_Order->__construct('1568')
#3 /app/public/wp-content/plugins/AffiliateWP-issue-2491/includes/admin/tools/upgrades/class-batch-upgrade-create-customers.php(117): Affiliate_WP_WooCommerce->get_customer('1568')
#4 /app/public/wp-content/plugins/AffiliateWP-issue-2491/includes/admin/ajax-actions.php(180): AffWP\Utils\Batch_Process\Upgrade_Create_Customers->process_step()
#5 /app/public/wp-includes/class-wp-hook.php(286): affwp_process_batch_request('')
#6 /app/public/wp-inclu in /app/public/wp-content/plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php on line 102

**Comment by pippinsplugins** [comment]
@gspice that error should now be addressed in `release/2.2`.

**Comment by tubiz** [comment]
I was able to replicate the error too on #2511. I have checked `release/2.2` and it hasn't been fixed in it.

https://github.com/AffiliateWP/AffiliateWP/blob/8aa1061bae132648ed76da009eed0edb1454650a/includes/customer-functions.php#L51-L59

**Comment by pippinsplugins** [comment]
Fixed!
